web/src/pages/api/admin/bullmq/index.ts

```diff
@@ -0,0 +1,128 @@
+import { type NextApiRequest, type NextApiResponse } from "next";
+import { z } from "zod";
+import {
+  BatchExportQueue,
+  IngestionQueue,
+  LegacyIngestionQueue,
+  logger,
+  QueueName,
+  TraceUpsertQueue,
+  DatasetRunItemUpsertQueue,
+  EvalExecutionQueue,
+} from "@langfuse/shared/src/server";
+import { env } from "@/src/env.mjs";
+import { type Queue } from "bullmq";
+
+/* 
+This API route is used by Langfuse Cloud to retry failed bullmq jobs.
+*/
+
+const RetryBullMqJobs = z.object({
+  action: z.literal("retry"),
+  queueNames: z.array(z.string()),
+});
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse,
+) {
+  try {
+    // allow only POST requests
+    if (req.method !== "POST") {
+      res.status(405).json({ error: "Method Not Allowed" });
+      return;
+    }
+
+    if (!env.NEXT_PUBLIC_LANGFUSE_CLOUD_REGION) {
+      res.status(403).json({ error: "Only accessible on Langfuse cloud" });
+      return;
+    }
+
+    // check if ADMIN_API_KEY is set
+    if (!env.ADMIN_API_KEY) {
+      logger.error("ADMIN_API_KEY is not set");
+      res.status(500).json({ error: "ADMIN_API_KEY is not set" });
+      return;
+    }
+
+    // check bearer token
+    const { authorization } = req.headers;
+    if (!authorization) {
+      res
+        .status(401)
+        .json({ error: "Unauthorized: No authorization header provided" });
+      return;
+    }
+    const [scheme, token] = authorization.split(" ");
+    if (scheme !== "Bearer" || !token || token !== env.ADMIN_API_KEY) {
+      res.status(401).json({ error: "Unauthorized: Invalid token" });
+      return;
+    }
+
+    const body = RetryBullMqJobs.safeParse(req.body);
+
+    if (!body.success) {
+      res.status(400).json({ error: body.error });
+      return;
+    }
+
+    if (body.data.action === "retry") {
+      logger.info(
+        `Retrying jobs for queues ${body.data.queueNames.join(", ")}`,
+      );
+
+      for (const queueName of body.data.queueNames) {
+        const queue = getQueue(queueName as QueueName);
+        let count = 0;
+        let failed;
+        let loopCount = 0;
+        const maxLoops = 100;
+
+        do {
+          if (loopCount >= maxLoops) {
+            logger.warn(
+              `Circuit breaker activated: Stopped after ${maxLoops} iterations for queue ${queueName}`,
+            );
+            break;
+          }
+
+          failed = await queue?.getJobs(["failed"], 0, 100, true);
+          logger.info(`Retrying jobs for queue ${JSON.stringify(failed)}`);
```

Comment:
What is in the `failed` object? Can it contain PII? In that case we should rather log the count or the queue name.

If it's just the job config, good to go.

Reply:
true. left over from testing. Removed.